drivers/SmartThings/matter-hrap/src/init.lua

```diff
@@ -0,0 +1,182 @@
+    device.log.info("Ssid is null. Per the spec, no primary Wi-Fi network is available at the moment.")
+    return
+  end
+  local valid_utf8, utf8_err = lustre_utils.validate_utf8(ib.data.value)
+  if valid_utf8 then
+    device:emit_event_for_endpoint(ib.endpoint, capabilities.wifiInformation.ssid({ value = ib.data.value }))
+  else
+    device.log.info("UTF-8 validation of Ssid failed: Error: '"..utf8_err.."'.")
+  end
+end
+
+local function thread_interface_enabled_attribute_handler(driver, device, ib)
+  if ib.data.value then
+    device:emit_event_for_endpoint(ib.endpoint, capabilities.threadBorderRouter.threadInterfaceState("enabled"))
+  else
+    device:emit_event_for_endpoint(ib.endpoint, capabilities.threadBorderRouter.threadInterfaceState("disabled"))
+  end
+end
+
+-- Spec uses TLV encoding of Thread Version, which should be mapped to a more user-friendly name
+local VERSION_TLV_MAP = {
+  [1] = "1.0.0",
+  [2] = "1.1.0",
+  [3] = "1.2.0",
+  [4] = "1.3.0",
+  [5] = "1.4.0",
+}
+
+local function thread_version_attribute_handler(driver, device, ib)
+  local version_name = VERSION_TLV_MAP[ib.data.value]
+  if version_name then
+    device:emit_event_for_endpoint(ib.endpoint, capabilities.threadBorderRouter.threadVersion({ value = version_name }))
+  end
+end
+
+
+--[[ COMMAND HANDLERS ]]--
+
+local threadNetwork = capabilities.threadNetwork
+local TLV_TYPE_ATTR_MAP = {
+  [0] = threadNetwork.channel,
+  [1] = threadNetwork.panId,
+  [2] = threadNetwork.extendedPanId,
+  [3] = threadNetwork.networkName,
+  [5] = threadNetwork.networkKey,
+}
+
+local function dataset_response_handler(driver, device, ib)
+  if ib.status ~= im.InteractionResponse.Status.SUCCESS then
+    log.error("Failed to retrieve thread operational dataset")
+    return
+  elseif not ib.info_block.data.elements.dataset then
+    log.debug("In dataset_response_handler, received an empty operational dataset")
+    return
+  end
+
+  local operational_dataset_length = ib.info_block.data.elements.dataset.byte_length
+  local spec_defined_max_dataset_length = 254
+  if operational_dataset_length > spec_defined_max_dataset_length then
+      log.error("In dataset_response_handler, operational dataset that was received is too long")
+      return
+  end
+
+  -- parse dataset
+  local operational_dataset = ib.info_block.data.elements.dataset.value
+  local cur_byte = 1
+  while cur_byte + 1 <= operational_dataset_length do
+    local tlv_type = string.byte(operational_dataset, cur_byte)
+    local tlv_length = string.byte(operational_dataset, cur_byte + 1)
+    if (cur_byte + 1 + tlv_length) > operational_dataset_length then
+      log.error("In dataset_response_handler, received a malformed operational dataaset")
+      return
+    end
+    local tlv_mapped_attr = TLV_TYPE_ATTR_MAP[tlv_type]
+    if tlv_mapped_attr then
+      local tlv_value = operational_dataset:sub(cur_byte + 2, cur_byte + 1 + tlv_length)
```

Comment:
```suggestion
      local tlv_value = operational_dataset:sub(cur_byte + 2, cur_byte + 2 + tlv_length)
```
Should this be '+2', or does the TLV length include the byte containing the TLV length?

Reply:
no, because the + 1 is the 1 byte tlv_length

Reply:
that's the `cur_byte`

Reply:
Gotcha, thanks!